tests/jballerina-unit-test/src/test/resources/test-src/typedefs/readonly-type-definitions-cyclic.bal

```diff
@@ -0,0 +1,33 @@
+// Copyright (c) 2022 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+type T readonly & (A|B);
+type A readonly & [T];
+type B readonly & [T];
+
+type C readonly & [D];
+type D readonly & [C];
+
+type T2 readonly & A2;
+type A2 readonly & [T2];
+
+function testReadonlyCyclicTypes() {
+    T? _ = ();
+    C? _ = ();
+    D? _ = ();
+    T2? _ = ();
+    A2? _ = ();
+}
```

Comment:
> Was there a reason to fix just this issue instead of also fixing
> 
> ```
> type T [T]
> ```
> 
> ?
> 
> May not be able to write some tests due to this and/or #38302.

This PR is mostly consider about readonly recursive types.
`type T [T]` is separatedly solved in https://github.com/ballerina-platform/ballerina-lang/pull/38276